### PR TITLE
Vectorize CPU ATen mean kernel for BF16 & FP16 dtypes

### PR DESCRIPTION
**Pull request by sanchitintel**
## Summary
Since #97351, CPU ATen kernel for `mean` for BF16 & FP16 dtypes has been unvectorized (it's not even implicitly vectorized).

This PR vectorizes `mean` for BF16 & FP16 on CPU in a `cast_fp32 -> sum -> div -> cast_bf16_or_fp16` fashion.

The perf benefit would be especially pronounced on machines with `AVX512_BF16` and/or `AVX512_FP16` ISA support.

## Benchmarking data for BF16 (collected before & after the change in this PR)

**Machine:** Intel&reg; Xeon&reg; (4th generation series, formerly codenamed Sapphire Rapids) Platinum 8468H
One socket (48 physical cores) - used `numactl --membind=0 --cpunodebind=0`
libtcmalloc & Intel OpenMP were preloaded

Environment variable used -
`KMP_AFFINITY=granularity=fine,compact,1,0 KMP_BLOCKTIME=1 KMP_SETTINGS=1 OMP_NUM_THREADS=48 MKL_NUM_THREADS=48`

**Workload:** E2E performance on BS 32 resnet50 (using BF16 via AMP) inference using oneDNN Graph JIT fuser (`mean` kernel is dispatched to eager mode ATen kernel, and is the bottleneck right now)

| **BEFORE:** Latency with unvectorized mean (lower is better)| **AFTER:** Latency with vectorized mean (lower is better)| Speedup due to vectorizing mean|
|----------------------------|-------------------------|------------|
|                19.1 ms           |                10.8  ms       | latency reduced by ~43.45%      |

**Benchmarking script for BF16 -**

 ```
import time
import torch
import torchvision

# enable oneDNN Graph JIT fuser
torch.jit.enable_onednn_fusion(True)
# AMP for JIT mode is enabled by default, and is divergent with its eager mode counterpart
torch._C._jit_set_autocast_mode(False)

# sample input should be of the same shape as expected inputs
example_input = torch.rand(32, 3, 224, 224)
# Using resnet50 from torchvision in this example for illustrative purposes,
# but the line below can indeed be modified to use custom models as well.
model = getattr(torchvision.models, "resnet50")().eval()

with torch.no_grad(), torch.cpu.amp.autocast(cache_enabled=False, dtype=torch.bfloat16):
    # Conv-BatchNorm folding for CNN-based Vision Models should be done with ``torch.fx.experimental.optimization.fuse`` when AMP is used
    import torch.fx.experimental.optimization as optimization
    # Please note that optimization.fuse need not be called when AMP is not used
    model = optimization.fuse(model)
    model = torch.jit.trace(model, (example_input))
    model = torch.jit.freeze(model)
    # a couple of warm-up runs
    model(example_input)
    model(example_input)
    # speedup would be observed in subsequent runs
    start = time.time()
    model(example_input)
    end = time.time()
    inference_time = (end - start) * 1000
    print("Inference time is ", inference_time)
```

cc @jgong5 @mingfeima @XiaobingSuper @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/ReduceOps.cpp
Comment (sanchitintel) [inline aten/src/ATen/native/ReduceOps.cpp:1349]: nit:

`output -> `input`

Will change it after all CI checks will complete

File: aten/src/ATen/native/ReduceOps.cpp
Comment (jgong5) [inline aten/src/ATen/native/ReduceOps.cpp:1372]: ditto

File: aten/src/ATen/native/ReduceOps.cpp
Comment (jgong5) [inline aten/src/ATen/native/ReduceOps.cpp:1367]: I guess you can share this line with fp32 path?

**Review by jgong5 — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/ReduceOps.cpp
Comment (sanchitintel) [reply aten/src/ATen/native/ReduceOps.cpp]: Oops, I had looked at the definition of `at::native::to` instead of `TensorBase::to`. Fixed it. Thank you!

File: aten/src/ATen/native/ReduceOps.cpp
Comment (sanchitintel) [reply aten/src/ATen/native/ReduceOps.cpp:1361]: No, because FP32 path is shared by some other dtypes as well. 
BTW, `mean_kernel_impl` is not being used by any dtype now.
Would you recommend removing it from the PyTorch codebase? Thanks!

Comment (sanchitintel) [reply aten/src/ATen/native/ReduceOps.cpp]: Eliminated duplicate code. Thank you!

**Review by jgong5 — APPROVED** [review]

**Comment by sanchitintel** [comment]
Hi @ezyang, can you please help review this PR? It requires approval from a core maintainer.
If you'd have availability to review, please ignore the numerous commits & please only look at the changes introduced in this PR.

Thank you! :)

**Comment by sanchitintel** [comment]
Hi @malfet, can you please help review this PR?
eager-mode `mean` kernel for BF16/FP16 datatypes has been unvectorized since March, and since it's called by `adaptive_avg_pool2d` in models such as resnet50, their performance with AMP is poor.
This PR fixes the issue by vectorizing `mean` computation for BF16 & FP16 dtypes as well.

Thanks!

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
LGTM

**Pull request closed without merging**
